framework/src/modules/token/cc_method.ts

```diff
@@ -22,182 +22,287 @@ import {
+	public constructor(stores: NamedRegistry, events: NamedRegistry, tokenMethod: TokenMethod) {
 		super(stores, events);
+		this._tokenMethod = tokenMethod;
 	}
 
 	public addDependencies(interoperabilityMethod: InteroperabilityMethod) {
 		this._interopMethod = interoperabilityMethod;
 	}
 
-	public async beforeApplyCCM(_ctx: BeforeApplyCCMsgMethodContext): Promise<void> {
-		// const { ccm } = ctx;
-		// const methodContext = ctx.getMethodContext();
-		// if (ccm.fee < BigInt(0)) {
-		// 	throw new Error('Fee must be greater or equal to zero.');
-		// }
-		// const { id: ownChainID } = await this._interopMethod.getOwnChainAccount(methodContext);
-		// const { messageFeeTokenID } = await this._interopMethod.getChannel(
-		// 	methodContext,
-		// 	ccm.sendingChainID,
-		// );
-		// const [feeTokenChainID, feeTokenLocalID] = splitTokenID(messageFeeTokenID);
-		// const userStore = this.stores.get(UserStore);
-		// if (!feeTokenChainID.equals(ownChainID)) {
-		// 	await userStore.addAvailableBalanceWithCreate(
-		// 		methodContext,
-		// 		ctx.trsSender,
-		// 		messageFeeTokenID,
-		// 		ccm.fee,
-		// 	);
-		// 	return;
-		// }
-		// const escrowStore = this.stores.get(EscrowStore);
-		// await escrowStore.deductEscrowAmountWithTerminate(
-		// 	methodContext,
-		// 	this._interopMethod,
-		// 	ccm.sendingChainID,
-		// 	feeTokenLocalID,
-		// 	ccm.fee,
-		// );
-		// const canonicalTokenID = await this._tokenMethod.getCanonicalTokenID(
-		// 	methodContext,
-		// 	messageFeeTokenID,
-		// );
-		// await userStore.addAvailableBalanceWithCreate(
-		// 	methodContext,
-		// 	ctx.trsSender,
-		// 	canonicalTokenID,
-		// 	ccm.fee,
-		// );
+	public async beforeCrossChainCommandExecution(ctx: BeforeApplyCCMsgMethodContext): Promise<void> {
+		const { trsSender, ccm } = ctx;
+		const methodContext = ctx.getMethodContext();
+		const relayerAddress = trsSender;
+		const { id: ownChainID } = await this._interopMethod.getOwnChainAccount(methodContext);
+		const { messageFeeTokenID } = await this._interopMethod.getChannel(
+			methodContext,
+			ccm.sendingChainID,
+		);
+		const { chainID: feeTokenChainID, localID: feeTokenLocalID } = messageFeeTokenID;
+		const userStore = this.stores.get(UserStore);
+
+		if (feeTokenChainID.equals(ownChainID)) {
+			const escrowedAmount = await this._tokenMethod.getEscrowedAmount(
+				methodContext,
+				ccm.sendingChainID,
+				feeTokenLocalID,
+			);
+			if (escrowedAmount < ccm.fee) {
+				this.events.get(BeforeCCCExecutionEvent).error(
+					methodContext,
+					{
+						sendingChainID: ccm.sendingChainID,
+						receivingChainID: ccm.receivingChainID,
+						messageFeeTokenID: feeTokenLocalID,
+						messageFee: ccm.fee,
+						relayerAddress,
+					},
+					TokenEventResult.FAIL_INSUFFICIENT_BALANCE,
+				);
+
+				throw new Error('Insufficient balance in the sending chain for the message fee.');
+			}
+
+			const escrowStore = this.stores.get(EscrowStore);
+			await escrowStore.deductEscrowAmountWithTerminate(
```

Comment:
here no need to terminate, i think it's only deducting the balance.
if we fetch the escrow account above when checking, you mutate the fetched value and set here. In this way, there will be one less encode/decode

Reply:
you could add additional function in escrow store to handle this case FYI